src/JsonApi/Serializer/ItemNormalizer.php

```diff
@@ -0,0 +1,599 @@
+                'The type of the "%s" attribute must be "array", "%s" given.', $attributeName, gettype($data)
+            ));
+        }
+
+        $collectionKeyType = $type->getCollectionKeyType();
+        $collectionKeyBuiltinType = null === $collectionKeyType ? null : $collectionKeyType->getBuiltinType();
+
+        $values = [];
+        foreach ($data as $rawIndex => $obj) {
+            $index = $rawIndex;
+
+            // Given JSON API forces ids to be strings, we might need to cast stuff
+            if (null !== $collectionKeyBuiltinType && 'int' === $collectionKeyBuiltinType) {
+                $index = (int) $index;
+            } elseif (null !== $collectionKeyBuiltinType && !call_user_func('is_'.$collectionKeyBuiltinType, $index)) {
+                throw new InvalidArgumentException(sprintf(
+                    'The type of the key "%s" must be "%s", "%s" given.',
+                    $index,
+                    $collectionKeyBuiltinType,
+                    gettype($index)
+                ));
+            }
+
+            $values[$index] = $this->denormalizeRelationFromArray($className, $obj, $context, $format);
+        }
+
+        return $values;
+    }
+
+    /**
+     * Gets JSON API components of the resource: attributes, relationships, meta and links.
+     *
+     * @param object      $object
+     * @param string|null $format
+     * @param array       $context
+     *
+     * @return array
+     */
+    private function getComponents($object, string $format = null, array $context)
+    {
+        if (isset($this->componentsCache[$context['cache_key']])) {
+            return $this->componentsCache[$context['cache_key']];
+        }
+
+        $attributes = parent::getAttributes($object, $format, $context);
+
+        $options = $this->getFactoryOptions($context);
+
+        $typeShortName = $className = '';
+
+        $components = [
+            'links' => [],
+            'relationships' => [],
+            'attributes' => [],
+            'meta' => [],
+        ];
+
+        foreach ($attributes as $attribute) {
+            $propertyMetadata = $this
+                ->propertyMetadataFactory
+                ->create($context['resource_class'], $attribute, $options);
+
+            $type = $propertyMetadata->getType();
+            $isOne = $isMany = false;
+
+            if (null !== $type) {
+                if ($type->isCollection()) {
+                    $valueType = $type->getCollectionValueType();
+
+                    $isMany = null !== $valueType && ($className = $valueType->getClassName()) && $this->resourceClassResolver->isResourceClass($className);
+                } else {
+                    $className = $type->getClassName();
+
+                    $isOne = null !== $className && $this->resourceClassResolver->isResourceClass($className);
+                }
+
+                $typeShortName = '';
+
+                if ($className && $this->resourceClassResolver->isResourceClass($className)) {
```

Comment:
Because `isResourceClass` is quite heavy (#1068) I'd prefer if we could use a variable to store it above instead of calling it a least twice.